File: ruby/playing_with_digits.rb
def dig_pow(n, p)
  target = n.digits.reverse.map.with_index { |dig, pow| dig**(pow + p) }.sum
  k = 1
  until k * n > target
    return k if k * n == target
    k += 1
  end
  -1
end

p dig_pow(89, 1)
p dig_pow(92, 1)
p dig_pow(695, 2)
p dig_pow(46288, 3)

File: go/playing_with_digits.go
package main

import (
	"fmt"
	"math"
	"strconv"
)

func sumDigPow(n, p int) int {
	pFloat := float64(p)
	nStr := strconv.Itoa(n)
	sum := 0.0
	for i, digStr := range nStr {
		digit, _ := strconv.Atoi(string(digStr))
		sum += math.Pow(float64(digit), pFloat+float64(i))
	}

	return int(sum)
}

func DigPow(n, p int) int {
	sum := sumDigPow(n, p)
	if sum%n == 0 {
		return sum / n
	}
	return -1
}

func main() {
	fmt.Println(DigPow(89, 1))
	fmt.Println(DigPow(92, 1))
	fmt.Println(DigPow(695, 2))
	fmt.Println(DigPow(46288, 3))
}
